### PR TITLE
[CRYS-766] fix tests/dir.phpt, add readme.md

### PR DESCRIPTION
**Pull request by dduca-sugarcrm**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -17,3 +17,15 @@ run-tests.php
...
+.buildpath
+.project
+tmp-php.ini
+tests/*.diff
+tests/*.exp
+tests/*.log
+tests/*.out
+tests/*.php
+tests/*.sh
+tests/instance/txt/sometext.txt
```
Comment (satanTime) [inline .gitignore]: I'm not sure that we need those changes.

File: sugarcrm/SugarShadow.php [hunk 2]
```diff
@@ -41,6 +41,9 @@ protected function createInstance($path)
 		$this->configCacheDir($path, dirname(__FILE__));
 		mkdir($path . '/custom', 0775, true);
 		$this->createSiTempate($path);
+		// createSiTempate not exist {{{
+		//$this->createSiTempate($path);
+		// }}}
```
Comment (satanTime) [inline sugarcrm/SugarShadow.php]: we don't need those commented lines.

File: sugarcrm/SugarShadow.php [hunk 3]
```diff
@@ -104,7 +107,7 @@ static function shadow($server){
 			die ('<h3>Invalid SugarCRM Instance</h3>');
 		}else{
 			if(!file_exists($info['path'])){
-				if($this->config['shadow']['createDir']) {
+				if($shadow->config['shadow']['createDir']) {
```
Comment (satanTime) [inline sugarcrm/SugarShadow.php]: good catch.

File: sugarcrm/shadow.ini [hunk 4]
```diff
@@ -1,3 +1,16 @@
 extension=shadow.so
-shadow.override=imagepng@w1,finfo_file@1,ext2mime@0,mime_content_type@0,ziparchive::open@w0,ziparchive::addfile@0,tempnam@w0
-shadow.mkdir_mask=0750
```
Comment (satanTime) [inline sugarcrm/shadow.ini:3]: I'm not sure that we need to change them.
I agree to add commented lines about all settings, but don't change them from default value.

File: readme.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+WIKI: <http://internalwiki.sjc.sugarcrm.pvt/index.php/Shadow_configuration>
+
+PHP module
+==========
+
+Shadow is implemented as PHP extension. Compile it using:
+
+```bash
+/path/to/php/install/bin/phpize
+./configure --with-php-config=//path/to/php/install/bin/php-config
```
Comment (satanTime) [inline readme.md]: `//`

File: readme.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+Set up Apache virtual hosts as follows:
+
+```bash
+<VirtualHost shadowmagic.com>
+   DocumentRoot /path/to/template/sugarcrm
+   ServerName shadow.com
+   ServerAlias *.shadow.com
+   RewriteEngine On
+   RewriteRule ^(/(cache|custom)/.*) /path/to/instances/%{SERVER_NAME}/$1
```
Comment (satanTime) [inline readme.md]: it's not a true. shadow has much more requirements for write rules. please check apache config on qa test shadow instance.

File: shadow.c [hunk 7]
```diff
@@ -1238,7 +1237,6 @@ static void shadow_is_writable(INTERNAL_FUNCTION_PARAMETERS)
 
 	ensure_dir_exists(instname, &shadow_wrapper, NULL TSRMLS_CC);
 	/* Check whether dir containing the file is writable */
-	zend_dirname(instname, strlen(instname));
```
Comment (satanTime) [inline shadow.c:1241]: what a reason for that change?

File: sugarcrm/SugarShadowWithoutMongo.php [hunk 8]
```diff
@@ -0,0 +1,101 @@
+<?php
```
Comment (satanTime) [inline sugarcrm/SugarShadowWithoutMongo.php]: we don't need that file for mothership.

File: sugarcrm/shadow.ini [hunk 9]
```diff
@@ -1,3 +1,18 @@
+; Enable Shadow extension module
 extension=shadow.so
-shadow.override=imagepng@w1,finfo_file@1,ext2mime@0,mime_content_type@0,ziparchive::open@w0,ziparchive::addfile@0,tempnam@w0
+
+; Determines if Shadow is enabled
+shadow.enabled=1
```
Comment (satanTime) [inline sugarcrm/shadow.ini]: I think we can comment all parameters.

File: sugarcrm/shadow.ini [hunk 10]
```diff
@@ -1,3 +1,18 @@
...
+
+; Override function
+;shadow.override=imagepng@w1,finfo_file@1,ext2mime@0,mime_content_type@0,ziparchive::open@w0,ziparchive::addfile@0,tempnam@w0
+
+; Shadow debug, 8191 - this number enable all debug options.
+; Messages send to std err or apache error.log.
+;shadow.debug=8191
+
+; Cache size for files or folders paths. 10000 is default.
+shadow.cache_size=10000
```
Comment (satanTime) [inline sugarcrm/shadow.ini]: new line is missed.

File: tests/chmod.phpt [hunk 11]
```diff
@@ -2,12 +2,14 @@
 Check chmod()
 --SKIPIF--
 <?php if (!extension_loaded("shadow")) print "skip"; ?>
+<?php if (exec('whoami')=='root') print "skip"; ?>
```
Comment (satanTime) [inline tests/chmod.phpt]: code standards. please add required spaces.

File: tests/chmod.phpt [hunk 12]
```diff
@@ -18,14 +20,14 @@ echo "Unshadow\n";
 var_dump(file_exists("$template/txt/new/dir/_twrite.txt"));
 var_dump(file_exists("$instance/txt/new/dir/_twrite.txt"));
 $s = stat("$instance/txt/new/dir/_twrite.txt");
-echo decoct($s["mode"] & 0777);
+echo substr(decoct($s["mode"]),-3);
```
Comment (satanTime) [inline tests/chmod.phpt]: code standards. please add required spaces.

File: tests/maindir.phpt [hunk 13]
```diff
@@ -4,10 +4,8 @@ Check directory listing for main dir
 <?php if (!extension_loaded("shadow")) print "skip"; ?>
 --FILE--
 <?php
-$template = "templatedir";
-$instance = "instance";
+require_once('setup.inc');
 chdir("instance");
-shadow($template, $instance, array("cache", "custom"));
```
Comment (satanTime) [inline tests/maindir.phpt:10]: why do we remove shadow initialization from here?

File: tests/override.phpt [hunk 14]
```diff
@@ -1,11 +1,12 @@
 --TEST--
 Check overrides
 --INI--
-shadow.override=imagepng@w1,finfo_file@1
+shadow.override=imagepng@w1,finfo_file@1,ext2mime@0,mime_content_type@0,ziparchive::open@w0,ziparchive::addfile@0,tempnam@w0
```
Comment (satanTime) [inline tests/override.phpt]: test is related only to `gd`, why do we need all other rewrites?

File: tests/readdir.phpt [hunk 15]
```diff
@@ -3,17 +3,22 @@ Check listing directory content
...
 --FILE--
-<?php 
+<?php
 require_once('setup.inc');
 
 function dirread($dir) {
-if ($dh = opendir($dir)) {
-	while (($file = readdir($dh)) !== false) {
-    	echo "filename: $file : filetype: " . filetype("$dir/$file") . "\n";
+    $aTmp=array();
```
Comment (satanTime) [inline tests/readdir.phpt]: spaces

File: tests/readdir.phpt [hunk 16]
```diff
@@ -3,17 +3,22 @@ Check listing directory content
...
+    if (($dh = dir($dir)) == true) {
+        while (($file = $dh->read()) !== false) {
+            $aTmp[]="filename: $file : filetype: " . filetype("$dir/$file") . "\n";
```
Comment (satanTime) [inline tests/readdir.phpt]: spaces

File: tests/readdir.phpt [hunk 17]
```diff
@@ -3,17 +3,22 @@ Check listing directory content
...
+<?php
 require_once('setup.inc');
 
 function dirread($dir) {
-if ($dh = opendir($dir)) {
-	while (($file = readdir($dh)) !== false) {
-    	echo "filename: $file : filetype: " . filetype("$dir/$file") . "\n";
+    $aTmp=array();
+    if (($dh = dir($dir)) == true) {
+        while (($file = $dh->read()) !== false) {
```
Comment (satanTime) [inline tests/readdir.phpt]: why did we change readdir to Directory object? We can provide extra test for that but we shouldn't change original logic.

File: tests/write_new_mask.phpt [hunk 18]
```diff
@@ -4,22 +4,27 @@ Check writing to files in new dir with dir mask
...
 --FILE--
-<?php 
+<?php
 require_once('setup.inc');
 
+mkdir("$template/txt/new/dir/");
+mkdir("$template/txt/new2/dir/");
+
 file_put_contents("$template/txt/new/dir/_twrite.txt", "writing as template\n");
-ini_set("shadow.mkdir_mask", 0700);
```
Comment (satanTime) [inline tests/write_new_mask.phpt:12]: it's a case of test, why do we remove that?

File: tests/write_new_mask.phpt [hunk 19]
```diff
@@ -4,22 +4,27 @@ Check writing to files in new dir with dir mask
...
 file_put_contents("$template/txt/new2/dir/_twrite.txt", "writing as template\n");
 
+clearstatcache();
+
 $stat = stat("$template/txt/new/dir");
-echo decoct($stat["mode"])."\n";
+echo substr(decoct($stat["mode"]),-5)."\n";
```
Comment (satanTime) [inline tests/write_new_mask.phpt]: spaces

File: shadow.c (hunk 7 above)
Comment (satanTime) [inline shadow.c:1241]: solved. removal is valid.

File: tests/dir.phpt [hunk 20]
```diff
@@ -3,15 +3,20 @@ Check listing directory content
...
 --FILE--
-<?php 
+<?php
 require_once('setup.inc');
 
 function dirread($dir) {
-if ($dh = dir($dir)) {
-	while (($file = $dh->read()) !== false) {
-    	echo "filename: $file : filetype: " . filetype("$dir/$file") . "\n";
+    $aTmp=array();
```
Comment (satanTime) [inline tests/dir.phpt]: spaces

File: tests/dir.phpt [hunk 21]
```diff
@@ -3,15 +3,20 @@ Check listing directory content
...
+    if (($dh = dir($dir)) == true) {
+        while (($file = $dh->read()) !== false) {
+            $aTmp[]="filename: $file : filetype: " . filetype("$dir/$file") . "\n";
```
Comment (satanTime) [inline tests/dir.phpt]: spaces

File: tests/write_new_mask.phpt [hunk 22]
```diff
@@ -4,22 +4,27 @@ Check writing to files in new dir with dir mask
 shadow.mkdir_mask = 0755
 --SKIPIF--
 <?php if (!extension_loaded("shadow")) print "skip"; ?>
+<?php if (exec('whoami')=='root') print "skip"; ?>
```
Comment (satanTime) [inline tests/write_new_mask.phpt]: spaces

File: shadow.c (hunk 7 above)
Comment (dduca-sugarcrm) [inline shadow.c:1241]: zend_dirname - do nothing

File: tests/maindir.phpt (hunk 13 above)
Comment (dduca-sugarcrm) [inline tests/maindir.phpt:10]: Shadow do initialization in setup.inc.

File: tests/shadow_settings.phpt [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+    print 'skip';
+} ?>
+--FILE--
+<?php
+require_once 'setup.inc';
+
+var_dump(shadow_get_config());
+
+$befor = ini_get_all('shadow');
+echo $befor['shadow.mkdir_mask']['local_value'].PHP_EOL;
```
Comment (satanTime) [inline tests/shadow_settings.phpt]: spaces between dot

File: tests/shadow_settings.phpt [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+ini_set('shadow.mkdir_mask', '0700');
+
+$after = ini_get_all('shadow');
+echo $after['shadow.mkdir_mask']['local_value'].PHP_EOL;
```
Comment (satanTime) [inline tests/shadow_settings.phpt]: spaces between dot

File: tests/write_new_mask.phpt [hunk 25]
```diff
@@ -3,36 +3,55 @@ Check writing to files in new dir with dir mask
...
+ini_set('shadow.mkdir_mask', '0700');
 file_put_contents("$template/txt/new2/dir/_twrite.txt", "writing as template\n");
+mkdir("$template/txt/new2/dir7000/");
+
+clearstatcache();
 
-$stat = stat("$template/txt/new/dir");
-echo decoct($stat["mode"])."\n";
+$stat = stat("$template/txt/new/dir/");
+echo substr(decoct($stat['mode']), -5)."\n";
```
Comment (satanTime) [inline tests/write_new_mask.phpt]: spaces between dot

Comment (satanTime) [inline tests/write_new_mask.phpt]: `PHP_EOL` ?

**Pull request merged**